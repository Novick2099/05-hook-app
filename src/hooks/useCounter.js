import { useState } from "react";

export const useCounter = ( initialState = 10 ) => {
    const [counter, setCounter] = useState(initialState);

    const increment = () => {
        setCounter(counter + 1);
    }
    const decrement = ()=>{
        setCounter(counter - 1);
    }
    const onReset = () => {
        setCounter( initialState );
    }
    return {
        counter,
        increment,
        decrement,
        onReset
    };
}