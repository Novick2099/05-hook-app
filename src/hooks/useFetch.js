import {useEffect, useRef, useState} from "react";

export const useFetch = ( url ) => {

    const isMounted = useRef(true);
    const [state, setState] = useState({data: null, loading: true, error: null});

    useEffect(()=>{
        return () => {
            isMounted.current = false;
        }
    },[])

    useEffect(()=>{

        setState({
            data: null,
            loading: true,
            error: null});

        fetch(url)
            .then(resp => resp.json())
            .then(data => {
                setTimeout(()=>{
                    //todo va con realExampleRef evita actual sobre el state de un componente que no esta montado
                    if(isMounted.current){
                        setState({
                            loading: false,
                            error: null,
                            //todo forma corta de eCmascript 6
                            data
                        })
                    }else{
                        console.log('Set state no se llamo');
                    }
                }, 2000);
            }).catch((error)=>{
            setState({
                    data: null,
                    loading: false,
                    error: 'CanÂ´t connect with API'
                });
        });
    },[url]);
    return state;
}