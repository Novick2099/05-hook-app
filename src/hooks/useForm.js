const {useState} = require("react");

export const useForm = (initialState = {}) => {

    //todo Argumentos posicionales
    const [values, setValues] = useState(initialState);

    const reset = ()=> {
      setValues( initialState );
    }

    //todo creacion de objeto usando desestructuracion
    const handleInputChange = ({ target })=>{
        setValues({
           ...values,
           [target.name]: target.value
        });
    }

    return [values, handleInputChange, reset];
}