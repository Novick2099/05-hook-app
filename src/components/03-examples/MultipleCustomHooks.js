import {useFetch} from "../../hooks/useFetch";
import {useCounter} from "../../hooks/useCounter";
import {Button} from "@material-ui/core";

const MultipleCustomHooks = () => {

    const {counter, increment} = useCounter(1);
    const {loading, data} = useFetch(`https://breakingbadapi.com/api/quotes/${counter}`);
    //todo !niega el null  a true y si se vuelve a negar regresara false
    const {author, quote} = !!data && data[0];
    

    return (
        <>
            <div>
                <h1>Breaking bdas Quotes</h1>
                <p>Counter{counter}</p>
            </div>
            { loading
                ?
                (<div data-id={"alert"}>
                    Loading..
                </div>) :
                (<div>
                        <p data-id={"author"}>{author}</p>
                        <p data-id={"quote"}>{quote}</p>
                </div>
                )
            }
            <Button onClick={increment}>
                Siguiente
            </Button>
        </>

    );
};

export default MultipleCustomHooks;
