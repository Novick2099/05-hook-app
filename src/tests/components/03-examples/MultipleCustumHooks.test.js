import React from 'react';
import MultipleCustomHooks from "../../../components/03-examples/MultipleCustomHooks";
import {shallow} from "enzyme";
import {useFetch} from "../../../hooks/useFetch";
import {useCounter} from "../../../hooks/useCounter";


jest.mock("../../../hooks/useCounter");
jest.mock("../../../hooks/useFetch");
//todo mocks a function (search in documentation)
describe('Pruebas en MultipleCustomHooks',()=>{

  useCounter.mockReturnValue({
    counter: 10,
    increment: jest.fn(),
  });

  it('Should match agains snapshot',() => {
// todo mock a return for the mocked function how it should be the first time the component renders
    useFetch.mockReturnValue({
        data: null,
      loading: true,
      error: null,
    })
    useCounter.mockReturnValue({
      counter: 10,
      increment: jest.fn(),
    });

    const wrapper = shallow(<MultipleCustomHooks/>);
    expect(wrapper).toMatchSnapshot();



  });

  it('should show information',() => {

    useCounter.mockReturnValue({
      counter: 10,
      increment: jest.fn(),
    });
    useFetch.mockReturnValue({
      data: [{
        author: 'Fernando Herrera',
        quote: 'Hola mundo'
      }],
      loading: false,
      error: 'Error',
    });
    const wrapper = shallow (<MultipleCustomHooks />);
    console.debug(wrapper.html());
    //los esta encontrando pero no los esta pasando ]??
    expect( wrapper.find("[data-id='alert']").exists()).toBe(false);
    expect( wrapper.find("[data-id='author']").text()).toBe('Fernando Herrera');
    expect( wrapper.find("[data-id='quote']").text()).toBe('Hola mundo')
  });

});