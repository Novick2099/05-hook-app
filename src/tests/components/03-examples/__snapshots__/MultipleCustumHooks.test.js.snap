// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en MultipleCustomHooks Should match agains snapshot 1`] = `
<Fragment>
  <div>
    <h1>
      Breaking bdas Quotes
    </h1>
    <p>
      Counter
      10
    </p>
  </div>
  <div
    data-id="alert"
  >
    Loading..
  </div>
  <WithStyles(ForwardRef(Button))
    onClick={[MockFunction]}
  >
    Siguiente
  </WithStyles(ForwardRef(Button))>
</Fragment>
`;
